src/compat/compat.h

```diff
@@ -25,10 +25,10 @@
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <sys/select.h>
-#include <sys/socket.h>
+#include <sys/socket.h> // IWYU pragma: export
```

Comment:
What symbols are these needed for (here and below), and what are the trade-offs between adding the pragma, and actually adding the include where it is used? We should probably have some rationale as to wether we are going to further include/"export" everything out of compat, or actually add includes to source files.

Reply:
> What symbols are these needed for (here and below)...

```
#include <sys/socket.h>                    // for socklen_t, size_t, ssize_t
```
```
#include <netinet/in.h>                    // for in6_addr, in_addr, s6_addr
```

Both headers are non-Windows.

> ... and what are the trade-offs between adding the pragma, and actually adding the include where it is used?

Platform-specific headers require to be guarded with `#ifdef ... #endif` macros. So adding them in every place, where they are required, is not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

> We should probably have some rationale as to wether we are going to further include/"export" everything out of compat, or actually add includes to source files.

My suggestion is to "export" platform-specific headers out of `compat.h` when IWYU points at them.

Reply:
> #include <sys/socket.h> .... size_t, ssize_t

Is this a bug? `size_t`/`ssize_t` do not come from `socket.h`.

> So adding them in every place, where they are required, is not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).

Isn't adding every header, everywhere it is used, the entire point of IWYU though?

This could also lead to having a single bloated compat.h, which when included everywhere, mitigates one of the other supposed benefit of IWYU, which is removing unneeded headers to reducing preprocessing time.

Reply:
It just seems more convenient to include compat.h or socks.h when using the `Sock` wrapper, instead of having to deal with low-level system includes everywhere `Sock` is used. I guess an alternative to the iwyu pragma would be to also wrap the low-level types, but not sure if this is actually better.

Reply:
> > #include <sys/socket.h> .... size_t, ssize_t
> 
> Is this a bug? `size_t`/`ssize_t` do not come from `socket.h`.

Perhaps, it is. But it is unrelated, as the `socket.h` header is required to provide the `socklen_t` symbol.

> > So adding them in every place, where they are required, is not very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself).
> 
> Isn't adding every header, everywhere it is used, the entire point of IWYU though?
> 
> This could also lead to having a single bloated compat.h, which when included everywhere, mitigates one of the other supposed benefit of IWYU, which is removing unneeded headers to reducing preprocessing time.

That is not what I meant. The `compat.h` header, which includes _platform-specific_ headers, allows to avoid repetition of code like that:
```c++
#ifdef WIN32
#include <winsock2.h>
#else
#include <sys/socket.h>
#endif
```